File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/cdk/integrations/base/context/AirbyteConnectorRunner.java
package io.airbyte.cdk.integrations.base.context;

import io.airbyte.cdk.integrations.base.config.ConnectorConfigurationPropertySource;
import io.micronaut.configuration.picocli.MicronautFactory;
import io.micronaut.context.ApplicationContext;
import io.micronaut.context.ApplicationContextBuilder;
import io.micronaut.context.env.CommandLinePropertySource;
import io.micronaut.context.env.Environment;
import picocli.CommandLine;

import static io.airbyte.cdk.integrations.base.JavaBaseConstants.ARGS_CATALOG_KEY;

public class AirbyteConnectorRunner {

    public static <R extends Runnable> void run(final Class<R> cls, final String... args) {
        ApplicationContextBuilder builder = buildApplicationContext(cls, args);
        try (ApplicationContext ctx = builder.start()) {
            run(cls, ctx, args);
        }
    }

    public static <R extends Runnable> void run(final Class<R> cls, final ApplicationContext ctx, final String... args) {
        final CommandLine commandLine = new CommandLine(cls, new MicronautFactory(ctx));
        commandLine.execute(args);
    }

    private static ApplicationContextBuilder buildApplicationContext(final Class<?> cls, final String[] args) {
        io.micronaut.core.cli.CommandLine commandLine = io.micronaut.core.cli.CommandLine.parse(args);
        ConnectorConfigurationPropertySource connectorConfigurationPropertySource = new ConnectorConfigurationPropertySource(commandLine);
        CommandLinePropertySource commandLinePropertySource = new CommandLinePropertySource(commandLine);
        return ApplicationContext
                .builder(cls, Environment.CLI)
                .propertySources(connectorConfigurationPropertySource, commandLinePropertySource);
    }
}

File: airbyte-integrations/connectors/source-oracle-v2/src/main/kotlin/io/airbyte/cdk/AirbyteConnectorRunner.kt
/*
 * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.cdk

import io.micronaut.core.cli.CommandLine as MicronautCommandLine
import io.airbyte.cdk.command.ConnectorCommandLinePropertySource
import io.airbyte.cdk.integrations.base.AirbyteTraceMessageUtility
import io.airbyte.cdk.integrations.util.ApmTraceUtils
import io.airbyte.cdk.integrations.util.ConnectorExceptionUtil
import io.github.oshai.kotlinlogging.KotlinLogging
import io.micronaut.configuration.picocli.MicronautFactory
import io.micronaut.context.ApplicationContext
import io.micronaut.context.ApplicationContextBuilder
import io.micronaut.context.env.CommandLinePropertySource
import io.micronaut.context.env.Environment
import picocli.CommandLine

/**
 * Replacement for the Micronaut CLI application runner that configures the CLI components and adds
 * the custom property source used to turn the arguments into configuration properties.
 */
class AirbyteConnectorRunner {

    enum class ConnectorType {
        SOURCE,
        DESTINATION
    }

    companion object {

        private val logger = KotlinLogging.logger {}

        @JvmStatic
        fun <R : Runnable> run(
            connectorType: ConnectorType,
            cls: Class<R>,
            vararg args: String,
        ) {
            val commandLine: MicronautCommandLine = MicronautCommandLine.parse(*args)
            val configPropertySource = ConnectorCommandLinePropertySource(commandLine)
            val commandLinePropertySource = CommandLinePropertySource(commandLine)
            val ctxBuilder: ApplicationContextBuilder =
                ApplicationContext.builder(cls, Environment.CLI, connectorType.name.lowercase())
                    .propertySources(configPropertySource, commandLinePropertySource)
            try {
                val ctx: ApplicationContext = ctxBuilder.start()
                run(cls, ctx, *args)
            } catch (e: Throwable) {
                logger.error(e) { "Unable to perform command." }
                // Many of the exceptions thrown are nested inside layers of RuntimeExceptions. An
                // attempt is made to find the root exception that corresponds to a configuration
                // error. If that does not exist, we just return the original exception.
                ApmTraceUtils.addExceptionToTrace(e)
                val rootThrowable = ConnectorExceptionUtil.getRootConfigError(Exception(e))
                val displayMessage = ConnectorExceptionUtil.getDisplayMessage(rootThrowable)
                // If the connector throws a config error, a trace message with the relevant
                // message should be surfaced.
                if (ConnectorExceptionUtil.isConfigError(rootThrowable)) {
                    AirbyteTraceMessageUtility.emitConfigErrorTrace(e, displayMessage)
                }
            }
        }

        @JvmStatic
        fun <R : Runnable> run(
            cls: Class<R>,
            ctx: ApplicationContext,
            vararg args: String,
        ) {
            CommandLine(cls, MicronautFactory(ctx)).execute(*args)
        }
    }
}
